libvirt_ap_passthrough: new test case to verify autostart of crypto device

smitterl opened the pull request:
With [s390-tools v2.22.0](https://github.com/ibm-s390-linux/s390-tools/releases/tag/v2.22.0) crypto passthrough device configurations are now easily persisted.
Add a test case to confirm that the configuration is restored at reboot and that therefore a mediated device is autostarted.

For this the code was refactored:
1. Move `MdevHandler` classes from existing virttools test case into new module under `provider` for re-use in new libvirt test case.
2. Create a new `ApMdevHandler`, refactor existing AP passthrough test case to use it instead.
3. Create new test case.

This change depends on https://github.com/avocado-framework/avocado-vt/pull/3472

smitterl commented:
Test runs to confirm test case passes and refactoring didn't introduce regression:
```
# avocado run --vt-type libvirt --vt-machine-type s390-virtio --vt-connect-uri qemu:///system libvirt_ap_passthrough,libvirt_ap_passthrough_autostart,libvirt_ccw_passthrough_read_write --test-runner=runner
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 10112f3bbde6e3242a91ff7957c0c23a9533471a
JOB LOG    : /var/lib/avocado/job-results/job-2022-07-29T12.57-10112f3/job.log
 (1/3) type_specific.io-github-autotest-libvirt.libvirt_ap_passthrough_autostart: PASS (53.19 s)
 (2/3) type_specific.io-github-autotest-libvirt.libvirt_ap_passthrough.hotplug: PASS (44.52 s)
 (3/3) type_specific.io-github-autotest-libvirt.libvirt_ap_passthrough.coldplug: PASS (24.60 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-07-29T12.57-10112f3/results.html
JOB TIME   : 122.71 s


# Needed to apply patch https://github.com/autotest/tp-libvirt/pull/4388
# avocado run --vt-type virttools --vt-machine-type s390-virtio --vt-connect-uri qemu:///system virt_install.hostdev.mdev --test-runner=runner
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 106d469b6eef98e95a1fb4a7a3f27ebecb62caf1
JOB LOG    : /var/lib/avocado/job-results/job-2022-07-29T13.03-106d469/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virt_install.hostdev.mdev.check_present_inside_guest: PASS (16.67 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-07-29T13.03-106d469/results.html
JOB TIME   : 17.05 s
```

smitterl commented:
@chunfuwen Please, can you help review?

chunfuwen commented on libvirt/tests/src/passthrough/ap/libvirt_ap_passthrough_autostart.py:
```diff
@@ -0,0 +1,77 @@
+    cmd = "mdevctl list -u %s" % uuid
+    err, out = cmd_status_output(cmd, shell=True, session=session)
+    if uuid in out:
+        pass
```
Do you intend to reserve some logic here? As per my understanding, in official release code, it is hard to find pass statement there

smitterl replied:
No, if the command prints the 'uuid' to stdout, then the device has been created correctly, so the code doesn't have to do anything.

chunfuwen commented on provider/vfio/ap.py:
```diff
@@ -0,0 +1,65 @@
+                        ignore_status=False)
+
+
+def create_autostart_configuration(domain_info, session=None):
```
the method name is create_autostart_configuration, so first impression to me is that it will return one config, but actually it return uuid. it sounds like the method name doesn't match with return value

smitterl replied:
@chunfuwen 
Would `set_up_autostart_configuration` a better name? Do you have another proposal?

chunfuwen replied:
How about create_autostart_mediated_device since "create" means that we need return something, and also returned 'uuid' is mediated device id

chunfuwen replied:
How about create_autostart_mediated_device since "create" means that we need return something, and also returned 'uuid' is mediated device id

smitterl replied:
Solved

chunfuwen reviewed (commented):
see inline comments

dzhengfy reviewed (commented):
Others LGTM

smitterl commented on libvirt/tests/src/passthrough/ap/libvirt_ap_passthrough_autostart.py:
```diff
@@ -0,0 +1,77 @@
+    if uuid in out:
+        pass
+    else:
+        raise TestFail("Mediated device UUID(%s) not listed in output:"
```
> Is below code simpler?
> 
> ```
> if uuid  not in out:
>     raise TestFail("Mediated device UUID(%s) not listed in output:"
> ```

@dzhengfy  I expect uuid to be listed in out and I don't have to do anything if that's the case; this is what the original code says. Your version is shorter but I think it's more complicated. If you prefer your version, please let me know. Thanks

smitterl replied:
Solved

smitterl commented on libvirt/tests/src/passthrough/ap/libvirt_ap_passthrough_autostart.py:
```diff
@@ -0,0 +1,77 @@
+from avocado.core.exceptions import TestFail
+
+from virttest.utils_misc import cmd_status_output
+from virttest.libvirt_xml.vm_xml import VMXML
```
Solved

smitterl commented on libvirt/tests/src/passthrough/ap/libvirt_ap_passthrough_autostart.py:
```diff
@@ -0,0 +1,77 @@
+from virttest.utils_zcrypt import load_vfio_ap
+
+from provider.vfio.mdev_handlers import MdevHandler
+from provider.vfio import ap
```
Solved

smitterl commented on provider/vfio/ap.py:
```diff
@@ -0,0 +1,65 @@
+    """
+
+    hostdev_xml = hostdev.Hostdev()
+    hostdev_xml.mode = "subsystem"
```
Solved

smitterl commented:
@dzhengfy @chunfuwen I addressed all your comments and re-ran the test cases. Please, can you help merge? Thank you!

```
# avocado run --vt-type libvirt --vt-machine-type s390-virtio --vt-connect-uri qemu:///system libvirt_ap_passthrough,libvirt_ap_passthrough_autostart,libvirt_ccw_passthrough_read_write --test-runner=runner
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : f81776e0fcf7e9b41c8644a1e7cc9a01f5d2030d
JOB LOG    : /var/lib/avocado/job-results/job-2022-08-23T04.28-f81776e/job.log
 (1/3) type_specific.io-github-autotest-libvirt.libvirt_ap_passthrough.hotplug: PASS (13.64 s)
 (2/3) type_specific.io-github-autotest-libvirt.libvirt_ap_passthrough.coldplug: PASS (29.73 s)
 (3/3) type_specific.io-github-autotest-libvirt.libvirt_ap_passthrough_autostart: PASS (56.47 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-08-23T04.28-f81776e/results.html
JOB TIME   : 100.23 s

# avocado run --vt-type virttools --vt-machine-type s390-virtio --vt-connect-uri qemu:///system virt_install.hostdev.mdev --test-runner=runner
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 350e46b42f6445a7fe563f48acf820f37b05316f
JOB LOG    : /var/lib/avocado/job-results/job-2022-08-23T04.30-350e46b/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virt_install.hostdev.mdev.check_present_inside_guest: PASS (20.18 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-08-23T04.30-350e46b/results.html
JOB TIME   : 20.52 s
```

dzhengfy reviewed (approved):
LGTM

chunfuwen reviewed (approved):
LGTM
